File: ContrailNet/ContrailNet/Contrail/IContrailSession.cs
using System;
using System.Collections.Generic;

namespace Contrail
{


	//
	// A ContrailSession is used to query an existing revision of the database or 
	// to create and commit a new revision.
	// 
	// <author>Ted Stockwell</author>
	//
	interface IContrailSession
	{
	
		// Get the database revision associated with this session. 
		long RevisionNumber { get; set; }
	
		// Create prepared query for items.
		IPreparedQuery<T> prepare<T> (ContrailQuery query) where T:Item;

		//
		// Commits all changes and closes the session
		// A client *MUST* call this method or the close method when finished with a session
		// No other methods may be invoked after invoking this method (except the 
		// close method, but calling the close method is not necessary after calling this method).
		// 
		// <throws>
		//	ConflictingCommitException when a potentially conficting change has been comitted to the database by another process.
		//  When this exception is throw the caller should open a new session and retry the transaction. 
		// <throws>
		//
		void commit ();

		//
		// Abandons any uncommitted changes and closes the session
		// A client *MUST* call this method or the commit method when finished with a transaction
		// No other methods may be invoked after invoking this method.
		// 
		// @throws ConflictingCommitException when a potentially conficting change has been comitted to the database by another process.
		// When this exception is throw the caller should open a new session and retry the transaction. 
		//
		void close ();

		boolean isActive ();
	
	
		// Return the raw storage
		IStorageProvider getStorageProvider ();
	
		bool create<E> (E entity) where E:Item;

		void store<E> (params E[] entities) where E:Item;

		void store<E> (IEnumerator<E> entities) where E:Item;

		void update<E> (params E[] entities) where E:Item;

		void update<E> (IEnumerator<E> entities) where E:Item;

		void delete (params Identifier[] paths);

		void delete (Collection<Identifier> paths);

		void delete<E> (params E[] entities);

		void delete<E> (IEnumerator<E> entities);

		E fetch<E> (Identifier path) where E:Item;

		ICollection<E> fetch<E> (params Identifier[] paths) where E:Item;

		ICollection<E> fetch<E> (Iterable<Identifier> paths) where E:Item;

		ICollection<Identifier> listChildren (Identifier path);

		IDictionary<Identifier, ICollection<Identifier>> listChildren (params Identifier[] paths);

		IDictionary<Identifier, ICollection<Identifier>> listChildren (ICollection<Identifier> paths);

		ICollection<Identifier> listChildren (E path);

		IDictionary<Identifier, ICollection<Identifier>> listChildren (params E[] paths);

		IDictionary<Identifier, ICollection<Identifier>> listChildren (Iterable<E> entities);

		ICollection<E> fetchChildren<E> (Identifier path) where E:Item;

		IDictionary<Identifier, ICollection<E>> fetchChildren<E> (params Identifier[] paths) where E:Item;

		IDictionary<Identifier, ICollection<E>> fetchChildren<E> (ICollection<Identifier> paths) where E:Item;

		IDictionary<Identifier, ICollection<C>> fetchChildren<E,C> (params E[] paths) where E:Item where C:Item;

		IDictionary<Identifier, ICollection<C>> fetchChildren<E,C> (IEnumerator<E> entities) where E:Item where C:Item;

		ICollection<C> fetchChildren<E,C> (E entity) where E:Item where C:Item;

		void deleteAllChildren (params Identifier[] paths);

		void deleteAllChildren (ICollection<Identifier> paths);

		void deleteAllChildren<E> (params E[] paths);

		void deleteAllChildren<E> (IEnumerator<E> entities);

	}

}

File: branches/scala/plugins/com.googlecode.contraildb.core/src/com/googlecode/contraildb/core/IContrailSession.java
package com.googlecode.contraildb.core;

import java.io.IOException;
import java.util.Collection;
import java.util.Map;

import com.googlecode.contraildb.core.storage.provider.IStorageProvider;




/**
 * A ContrailSession is used to query an existing revision of the database or 
 * to create and commit a new revision.
 * 
 * @author Ted Stockwell
 */
public interface IContrailSession
{
	
	/**
	 * Get the database revision associated with this session. 
	 */
	public long getRevisionNumber();
	
	/**
	 * Query for items.
	 */
	public <T extends Item> IPreparedQuery<T> prepare(ContrailQuery query) throws IOException;

	/**
	 * Commits all changes and closes the session
	 * A client *MUST* call this method or the close method when finished with a session
	 * No other methods may be invoked after invoking this method (except the 
	 * close method, but calling the close method is not necessary after calling this method).
	 * 
	 * @throws ConflictingCommitException when a potentially conficting change has been comitted to the database by another process.
	 * When this exception is throw the caller should open a new session and retry the transaction. 
	 */
	public void commit() throws IOException, ConflictingCommitException;

	/**
	 * Abandons any uncommitted changes and closes the session
	 * A client *MUST* call this method or the commit method when finished with a transaction
	 * No other methods may be invoked after invoking this method.
	 * 
	 * @throws ConflictingCommitException when a potentially conficting change has been comitted to the database by another process.
	 * When this exception is throw the caller should open a new session and retry the transaction. 
	 */
	public void close() throws IOException;

	public boolean isActive();
	
	
	/**
	 * Return the under raw storage
	 */
	public IStorageProvider getStorageProvider();
	
	
	
	public <E extends Item> boolean create(E entity) throws IOException;
	public <E extends Item> void store(E... entities) throws IOException;
	public <E extends Item> void store(Iterable<E> entities) throws IOException;
	public <E extends Item> void update(E... entities) throws IOException;
	public <E extends Item> void update(Iterable<E> entities) throws IOException;

	public void delete(Identifier... paths) throws IOException;
	public void delete(Collection<Identifier> paths) throws IOException;
	public <E extends Item> void delete(E... entities) throws IOException;
	public <E extends Item> void delete(Iterable<E> entities) throws IOException;

	public <E extends Item> E fetch(Identifier path) throws IOException;
	public <E extends Item> Collection<E> fetch(Identifier... paths) throws IOException;
	public <E extends Item> Collection<E> fetch(Iterable<Identifier> paths) throws IOException;

	public Collection<Identifier> listChildren(Identifier path) throws IOException;
	public Map<Identifier, Collection<Identifier>> listChildren(Identifier... paths) throws IOException;
	public Map<Identifier, Collection<Identifier>> listChildren(Collection<Identifier> paths) throws IOException;
	public <E extends Item> Collection<Identifier> listChildren(E path) throws IOException;
	public <E extends Item> Map<Identifier, Collection<Identifier>> listChildren(E... paths) throws IOException;
	public <E extends Item> Map<Identifier, Collection<Identifier>> listChildren(Iterable<E> entities) throws IOException;

	public <E extends Item> Collection<E> fetchChildren(Identifier path) throws IOException;
	public <E extends Item> Map<Identifier, Collection<E>> fetchChildren(Identifier... paths) throws IOException; 
	public <E extends Item>  Map<Identifier, Collection<E>> fetchChildren(Collection<Identifier> paths) throws IOException; 
	public <E extends Item, C extends Item> Map<Identifier, Collection<C>> fetchChildren(E... paths) throws IOException; 
	public <E extends Item, C extends Item> Map<Identifier, Collection<C>> fetchChildren(Iterable<E> entities) throws IOException; 
	public <E extends Item, C extends Item> Collection<C> fetchChildren(E entity) throws IOException; 

	public void deleteAllChildren(Identifier... paths) throws IOException;
	public void deleteAllChildren(Collection<Identifier> paths) throws IOException;
	public <E extends Item> void deleteAllChildren(E... paths) throws IOException;
	public <E extends Item> void deleteAllChildren(Iterable<E> entities) throws IOException;
	


}
